Add new tutorial for deciphering viral populations using SNV and baculovirus isolates (Variant analysis)

wennj opened the pull request:
Dear GTN Team,

I am a new contributor and have created a tutorial for analysing virus populations using SNV (and using the example of large dsDNA viruses). The workflow is based on published results and follows a scientific approach. I have followed the GTN instructions as closely as possible (and as I understand them). Please excuse any mistakes in using Github. The tutorial has also been reviewed by a student.

Key changes:
- Added a new step-by-step tutorial (variant-analysis/baculovirus-isolate-variation) for analysing viral populations on the example of baculovirus isolates
- CONTRIBUTORS.yaml was changed.

Checked:
- Images can be re-hosted. 

Help:
- One image in the introduction is linked to ICTV.global. Allowed?

TODO:
- Tool bcftools mpileup / bcftools call are using DPR (deprecated but functional). Needs to be changed to AD in the future.
- Awaiting feedback and suggestions for improvement.
- Refine the tutorial based on reviewer comments.

Best,
JTW

bgruening commented on Gemfile.lock:
Please do not change this file.

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+objectives:
+- Understand the significance of single nucleotide variants (SNVs) in baculovirus populations.
+- Determine variable SNV positions for multiple isolates using a common reference genome.
+- Transform the output (VCF file) to a readable table format using tools available at Galaxy only. 
+- Interpret the SNV data to analysis the intra-isolate variability of baculovirus isolates or samples.
```
```suggestion
- Transform the output (VCF file) to a readable table format
- Interpret the SNV data to analyse the intra-isolate variability of baculovirus isolates or samples.
```

Not sure if the transform step is really an objective.

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+    SRA: Sequence Read Archive
+key_points:
+- Baculovirus populations are heterogeneous and show genetic variability.
+- Some SNV positions occur only in certain isolates and are therefore specific for that isolate.
```
```suggestion
- Some SNV positions occur only in certain isolates and are therefore specific to that isolate.
```

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+> The genome of the isolate Autographa californica multiple nucleopolyhedrovirus isolate C6 (AcMNPV-C6)
+> (family *Baculoviridae*, genus *Alphabaculovirus*, species *Alphabaculovirus aucalifornicae*) is one of the 
+> best-studied baculovirus genomes and is 133,894 bp long ({% cite Ayres1994 %}). 
+> It is the first fully sequenced genon of baculoviruses and today hundreds of genomes are fully sequenced
```
```suggestion
> It is the first fully sequenced genome of baculoviruses and today hundreds of genomes are fully sequenced
```

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+>
+{: .comment}
+
+The genome size makes it difficult to study genetic variation within a baculovirus population, 
+especially since the most commonly used sequencing technique (Illumina sequencing) 
+generates only short reads and requires genome assembly. Since genome assembly generates 
+a consensus sequence that reflects the majority of a sequenced virus population, 
+occuring genetic variation is masked. Tools for haplotype-sensitive assembly are 
+available but so far have been establisehd for viruses with a relatively short genome 
+but not for baculoviruses or other large dsDNA viruses (REFERENCES). Using Nanopore sequencing, 
+it is possible to sequence signficiant fragments of  baculovirus genomes to determine major 
+haplotypes ({% cite Wennmann2024 %}). 
```
```suggestion
The genome size makes it difficult to study genetic variation within a baculovirus population, 
especially since the most commonly used sequencing technique (Illumina sequencing) 
generates only short reads and requires genome assembly. Since genome assembly generates 
a consensus sequence that reflects the majority of a sequenced virus population, 
occurring genetic variation is masked. Tools for haplotype-sensitive assembly are 
available but so far have been established for viruses with a relatively short genome 
but not for baculoviruses or other large dsDNA viruses (REFERENCES). Using Nanopore sequencing, 
it is possible to sequence significant fragments of baculovirus genomes to determine major 
haplotypes ({% cite Wennmann2024 %}). 
```

wennj replied:
Suggestions were changed manually.

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+it is possible to sequence signficiant fragments of  baculovirus genomes to determine major 
+haplotypes ({% cite Wennmann2024 %}). 
+
+Single nucleotide variants (SNV) have been proven as a powerful tool for analyzing 
+the genetic variability of sequenced baculovirus isolates ({% cite Fan2020 %}). SNV 
+are particularly useful for the analysis of intra-isolate or intra-sample specific 
+variation, as a many bioinformatic workflows and tools are established for SNV determination 
+and processing. For the identification of SNVs, Illumina sequencing data is mostly used, 
+as this sequencing technique provides highly accurate reads with a very low probability 
+of sequencing error. The data is usually provided and stored in fatsq or fastqsanger format.
```
```suggestion
Single nucleotide variants (SNV) have been proven as a powerful tool for analyzing 
the genetic variability of sequenced baculovirus isolates ({% cite Fan2020 %}). SNV 
are particularly useful for the analysis of intra-isolate or intra-sample specific 
variation, as many bioinformatic workflows and tools are established for SNV determination 
and processing. For the identification of SNVs, Illumina sequencing data is mostly used, 
as this sequencing technique provides highly accurate reads with a very low probability 
of sequencing error. The data is usually provided and stored in FASTQ or FASTQsanger format.
```

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+as this sequencing technique provides highly accurate reads with a very low probability 
+of sequencing error. The data is usually provided and stored in fatsq or fastqsanger format.
+
+When a baculovirus sample or isolate is sequenced, a dataset of sequence reads is 
+obtained that consists of a multitude of fragmented genomes and the excact reconstruction 
+of individual whole genomes is no longer possible. However, variable SNV positions can 
+be used as markers to count the frequency of the nucleotides that occur in the SNV positions. 
+The determination of the nucleotides is only possible because not one but many genomes 
+were sequenced. Therefor, the SNVs reflect the genetic variability of the virus population itself. 
```
```suggestion
When a baculovirus sample or isolate is sequenced, a dataset of sequence reads is 
obtained that consists of a multitude of fragmented genomes and the exact reconstruction 
of individual whole genomes is no longer possible. However, variable SNV positions can 
be used as markers to count the frequency of the nucleotides that occur in the SNV positions. 
The determination of the nucleotides is only possible because not one but many genomes 
were sequenced. Therefore, the SNVs reflect the genetic variability of the virus population itself. 
```

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+The workflow is based on sequencing data from the Cydia pomonella granulovirus (CpGV) 
+(family *Baculoviridae*, genus *Betabaculovirus*). CpGV is used commercially in agriculture 
+for the biological control the larvae of the codling moth (*Cydia pomonella*), which cause 
+significant losses to  fruit crops, such als apple and pear.
```
```suggestion
significant losses to fruit crops, such as apples and pears.
```

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+
+## Reference Genome from NCBI Genbank
+
+Selecting the right or most suitable reference sequence is important if not critical for the entire analysis. The referen genome used in this tutorial is the Cydia pomonella granulovirus (Mexican isolate, CpGV-M). The function of the reference is to act as an anchor point to link detected SNV position in all sequenced and analyzed isolates. Selecting a correct reference genome is critical, becaus ...
```
```suggestion
Selecting the right or most suitable reference sequence is important if not critical for the entire analysis. The reference genome used in this tutorial is the Cydia pomonella granulovirus (Mexican isolate, CpGV-M). The function of the reference is to act as an anchor point to link detected SNV positions in all sequenced and analyzed isolates. Selecting a correct reference genome is critical because it must be as closely related as possible to the sequenced samples/isolates. Therefore, it is best to use a reference from the same species that has been well analyzed by previous studies or the scientific community.  The following step will download the NCBI Accession Number `KM217575`, which corresponds to the `Cydia pomonella granulovirus isolate CpGV-M, complete genome`. It will later serve as the reference genome for the detection of SNV positions.  
```

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+>
+{: .hands_on}
+
+## Paired-end Sequencing Data from NCBI SRA
+
+The tutorial is based on Illumina data sets from several isolates of the Cydia pomonella granulovirus (CpGV). 
+The CpGV is one of the most well studied baculoviruses because many isolates have been sequenced and sequence 
+data sets are available at NCBI Genbank and NCBI SRA. We will use four CpGV isolates to decipher their 
+intra-isolate specific variation and try to draw conclusios about the isolate's homogenity or heterogenity. 
+We will also encounter a mixed isolate (a more or less clean mixture of other previously sequenced CpGV isoaltes) 
+and learn how to recognize it. In addition, we will learn how to determine the composition 
+of this mixted isolate based on other sequenced CpGV isolates.
+
```
```suggestion
## Paired-end Sequencing Data from NCBI SRA

The tutorial is based on Illumina data sets from several isolates of the Cydia pomonella granulovirus (CpGV). 
The CpGV is one of the most well-studied baculoviruses because many isolates have been sequenced and sequence 
data sets are available at NCBI Genbank and NCBI SRA. We will use four CpGV isolates to decipher their 
intra-isolate specific variation and try to draw conclusions about the isolate's homogeneity or heterogeneity. 
We will also encounter a mixed isolate (a more or less clean mixture of other previously sequenced CpGV isolates) 
and learn how to recognize it. In addition, we will learn how to determine the composition 
of this mixed isolate based on other sequenced CpGV isolates.

```

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+# Read Preprocessing and Quality Control
+
+After the reference genome and the paired-end Illumina reads are loaded in the history, 
+it is time to start the analysis. First, the reads needs to be adapter trimmed and
```
```suggestion
it is time to start the analysis. First, the reads need to be adapter trimmed and
```

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+
+> <tip-title>Learn more about quality of reads!</tip-title>
+> Before and after trimming, it is usually a good idea to check the quality of the reads using FASTQC. 
+> Howver, we will omit this step here. 
```
```suggestion
> However, we will omit this step here. 
```

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+>
+{: .hands_on}
+
+What you get is a file in Variant Call Format (VCF), which can be difficult to understand at first glance. It is important to note that the file begins with a large header, lines that begin with two hashtags (##), which contain details of the analysis and the data set. After the header, there is a table with a line that contains the column names. This line begins with just one hashtag (#). After  ...
```
```suggestion
What you get is a file in Variant Call Format (VCF), which can be difficult to understand at first glance. It is important to note that the file begins with a large header, lines that begin with two hashtags (##), which contain details of the analysis and the data set. After the header, there is a table with a line that contains the column names. This line begins with just one hashtag (#). After that, the results are displayed in a tab-separated format. In our VCF example, some columns have the following abbreviations and meanings (I just selected a few that are important to us in this tutorial):
```

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+|**DPR**   |The read depth for each allele. Here, the first value corresponds to the reference nucleotide (REF), the second to the first possible allele (ALT1), the second to the second possible allele (ALT2) and the third value to the last possible allele (ALT3).
+
+> <comment-title>DPR functional, but deprecated</comment-title>
+> If you start **bcftools mpileup** with the *Optional tags to output* `DPR (Number of high quality bases for each observed allele)` option, you will see a warning in the information panel of the VCF file: `[warning] tag DPR functional, but deprecated. Please switch to AD in future`. At the time of writing this tutorial, I was using DPR. It works the same way with *Optional tags to output* option ...
```
```suggestion
> If you start **bcftools mpileup** with the *Optional tags to output* `DPR (Number of high-quality bases for each observed allele)` option, you will see a warning in the information panel of the VCF file: `[warning] tag DPR functional but deprecated. Please switch to AD in future. At the time of writing this tutorial, I was using DPR. It works the same way with *Optional tags to output* option `AD` (instead of DPR), but you will need to adjust something later in the workflow. The tutorial will be switched to AD in the near future. For now, we stick with DPR.
```

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+>    - Click Run Tool
+>
+>    > <comment-title> Column names and numbers </comment-title>
+>    > To provide ggplot2 with the data for the X- and Y-axis, column numbers must be passed to the function. You have to check which column numbers the columns `POS` and `REL.ALT` have. In addition, ggplot2 offers the option of to split the data according to `SAMPLE`, to obtain a separate plot for each value in SAMPLE. 
```
```suggestion
>    > To provide ggplot2 with the data for the X- and Y-axis, column numbers must be passed to the function. You have to check which column numbers the columns `POS` and `REL.ALT` have. In addition, ggplot2 offers the option of splitting the data according to `SAMPLE`, to obtain a separate plot for each value in SAMPLE. 
```

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+
+# SNV Specificity Determination
+
+Now we come to the last but most complex section of this tutorial - determining SNV specificities. We have seen that CpGV-V15 is a mixed isolate. Now we want to find out which isolates were mixed and how this SNV pattern can be explained. In the next step, we will consider  SNV positions as markers  and look for SNV positions that are only variable for one or more sequenced isolates of CpGV. If a ...
```
```suggestion
Now we come to the last but most complex section of this tutorial - determining SNV specificities. We have seen that CpGV-V15 is a mixed isolate. Now we want to find out which isolates were mixed and how this SNV pattern can be explained. In the next step, we will consider  SNV positions as markers and look for SNV positions that are only variable for one or more sequenced isolates of CpGV. If an SNV position is only variable for CpGV-S (i.e. the relative frequency is greater than 0; `REL.ALT > 0`) but is equal to zero for the other isolates, then this position is a marker for CpGV-S, or specific for CpGV-S. Since we cannot perform this determination by hand for all positions, I have written a small program implemented in the **Text reformatting with awk** tool that does the work for us. 
```

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+
+## Replace SRA Names with Virus Abbreviations
+
+One thing that stands out are the SAMPLE names, which were taken automatically from the NCBI SRA datasets. Since it is difficult to remember which virus isolate is behind which SRA number, we can replace the accession numbers with proper names. This makes the table even easier to read and later we can use the information directly to display the SNV positions. 
```
```suggestion
One thing that stands out is the SAMPLE names, which were taken automatically from the NCBI SRA datasets. Since it is difficult to remember which virus isolate is behind which SRA number, we can replace the accession numbers with proper names. This makes the table even easier to read and later we can use the information directly to display the SNV positions. 
```

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+
+# Visualizing SNV Variability Across Isolates
+
+The final table should be much easier to read and contain all the information we need to perform an analysis of the intra-isolate specific variability. Here is an short overview about the final table (selected columns) and its first three SNV positions:  
```
```suggestion
The final table should be much easier to read and contain all the information we need to perform an analysis of the intra-isolate specific variability. Here is a short overview of the final table (selected columns) and its first three SNV positions:  
```

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+
+![SNV specificity plot for CpGV-V15](../../images/baculovirus-isolate-variation/SNV_specificity_plot_CpGV-V15.png  "Deciphering the composition of the isolate CpGV-V15 by using SNV specificities.")
+
+Most of the SNV positions specific for CpGV-E2 have a frequency around 0.5 (50%) (Fig. 5, SNV specificity: CpGV-E2). Looking at the positions specific for CpGV-S, we also see a SNV cloud at around 0.5 (50 %) (Fig. 5, SNV specificity: CpGV-S). From these two observations, we can conclude that CpGV-E2 and CpGV-S occur at a ratio of 50% each. As a control, we look at the markers (SNV positions) that ...
```
```suggestion
Most of the SNV positions specific for CpGV-E2 have a frequency of around 0.5 (50%) (Fig. 5, SNV specificity: CpGV-E2). Looking at the positions specific to CpGV-S, we also see a SNV cloud at around 0.5 (50 %) (Fig. 5, SNV specificity: CpGV-S). From these two observations, we can conclude that CpGV-E2 and CpGV-S occur at a ratio of 50% each. As a control, we look at the markers (SNV positions) that are specific for both CpGV-E2 and CpGV-S (Fig. 5, SNV specificity: CpGV-E2 + CpGV-S). Here we see that most SNV positions have a relative frequency of ~1 (100%). This also makes sense, because if CpGV-V15 is a mixture of CpGV-E2 and CpGV-S, then the markers that indicate both isolates should be at 100%, which is the case. Based on the SNV analysis, it can be concluded that CpGV-V15 is mainly a mixture of CpGV-S and CpGV-E2.  
```

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+- Baculovirus populations are heterogeneous and show genetic variability.
+- Some SNV positions occur only in certain isolates and are therefore specific for that isolate.
+- SNV specificities can be used as markers to identify isolates and determine mixtures.
+contributors:
```
You can extend this for example by:

contributions:
  authorship:
    - wennj
  editing:
    - your-student

wennj replied:
She is not on Github.

bgruening commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+layout: tutorial_hands_on
+
+title: Deciphering Virus Populations - Single Nucleotide Variants (SNVs) and Specificities in Baculovirus Isolates
+subtopic: ''
```
```suggestion
subtopic: 'one-health'
```

Maybe?

wennj replied:
Yes it is ok, although I think that the One Heatlh description ("Variant calling for monitoring and preventing health threats") is not very fitting.

bgruening reviewed (commented):
Very cool! Welcome @wennj! 

I have added a few batch of comments, if you agree with them you can accept them from inside the github interface.

wennj commented:
@bgruening Many thanks for the suggestions and feedback. I have addressed all comments. As this is my first tutorial and my first pull request, I hope everything is checked correctly.

bgruening commented:
Thanks a lot! Let's wait for some experts in this area to get some content feedback.

Can you please add the `jki` to this file as well: https://github.com/galaxyproject/training-material/blob/main/ORGANISATIONS.yaml this will then generate a page like: https://training.galaxyproject.org/hall-of-fame/uni-freiburg/

wennj commented:
> Thanks a lot! Let's wait for some experts in this area to get some content feedback.
> 
> Can you please add the `jki` to this file as well: https://github.com/galaxyproject/training-material/blob/main/ORGANISATIONS.yaml this will then generate a page like: https://training.galaxyproject.org/hall-of-fame/uni-freiburg/

I would like to check that first with my administration. Can this also be done at a later stage?

bgruening commented on CONTRIBUTORS.yaml:
```diff
@@ -1148,6 +1148,19 @@ jfallmann:
+      country: DE
+      lat: 49.449466
+      lon: 8.640339
+    affiliations:
+      - jki
```
```suggestion
```

You can remove this until you have permissions.

wm75 commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/workflows/main_workflow.ga:
```diff
@@ -0,0 +1 @@
+{"a_galaxy_workflow": "true", "annotation": "This standard workflow determines SNV positions from baculovirus isolates using bcftools.", "comments": [{"child_steps": [8, 9, 10, 11], "color": "pink", "data": {"title": "VCF to table transformation"}, "id": 4, "position": [2052.7, 226.2], "size": [794, 405], "type": "frame"}, {"child_steps": [14, 15, 16, 17, 18, 19], "color": "none", "data": {"title ...
```
Can you run this file through:
`python3 -m json.tool main_workflow.ga > main_workflow_formatted.ga`
to make it multi-line with standard indentation please?
This will greatly facilitate reviewing of diffs later.

wennj replied:
Thank you. very much for the help. I thought something was wrong all the time, but I didn't know how to solve it (and why it happened at all).

wm75 commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+
+# Introduction
+
+Baculoviruses of the family *Baculoviridae* ({% cite Harrison2018 %}) are one of the most intensively 
```
```suggestion
Baculoviruses of the family *Baculoviridae* ({% cite Harrison2018 %}) are among the most intensively 
```

wm75 commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+Baculoviruses of the family *Baculoviridae* ({% cite Harrison2018 %}) are one of the most intensively 
+studied viruses, not only because of their widespread application in biotechnology, medicine, and 
+agriculture ({% cite vanOers2007 %}). Like all members of the class *Naldaviricetes*, 
+baculoviruses have a circular large dsDNA genome that, in the family *Baculoviridae*, 
+ranges between 90 to 180 kbp and can encode up to 180 open reading frames (ORF). 
```
```suggestion
baculoviruses have large circular dsDNA genomes that, in the family *Baculoviridae*, 
range from 90 to 180 kbp and can encode up to 180 open reading frames (ORFs). 
```

wm75 commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+
+When multiple baculovirus isolates have been sequenced, SNV positions can be determined 
+that are specific to only certain isolates. These specific SNV markers can then be used 
+to identify and isolate based on sequencing data only, even when the isolate is present 
```
```suggestion
to identify an isolate based on sequencing data only, even when the isolate is present 
```

wm75 commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+
+The workflow is based on sequencing data from the Cydia pomonella granulovirus (CpGV) 
+(family *Baculoviridae*, genus *Betabaculovirus*). CpGV is used commercially in agriculture 
+for the biological control the larvae of the codling moth (*Cydia pomonella*), which cause 
```
```suggestion
for the biological control of larvae of the codling moth (*Cydia pomonella*), which cause 
```

wm75 commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+
+> <hands-on-title>NCBI SRA Data Download</hands-on-title>
+>
+>    Run {% tool [Download and Extract Reads in FASTAQ format from NCBI SRA](toolshed.g2.bx.psu.edu/repos/iuc/sra_tools/fastq_dump/3.1.1+galaxy1) %} with the following parameters:
```
```suggestion
>    Run {% tool [Download and Extract Reads in FASTQ format from NCBI SRA](toolshed.g2.bx.psu.edu/repos/iuc/sra_tools/fastq_dump/3.1.1+galaxy1) %} with the following parameters:
```

wm75 commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+>
+> 1. {% tool [Trim Galore!](toolshed.g2.bx.psu.edu/repos/bgruening/trim_galore/trim_galore/0.6.7+galaxy0) %} with the following parameters:
+>    - *"Is this library paired- or single-end?"*: `Paired Collection`
+>        - {% icon param-collection %} *"Select a paired collection"*: `output` (output of **Download and Extract Reads in FASTAQ format from NCBI SRA** {% icon tool %})
```
```suggestion
>        - {% icon param-collection %} *"Select a paired collection"*: `Paired-end data (fastq-dump)` (output of **Download and Extract Reads in FASTAQ format from NCBI SRA** {% icon tool %})
```

In general,  in the GTN we are not using a tool's output name (as defined in the tool wrapper), but the user facing label of the generated dataset if that is a nice one, or a label-like description of the dataset. Take a look at https://training.galaxyproject.org/training-material/topics/evolution/tutorials/mtb_transmission/tutorial.html as an example, then change most of your hands-on boxes accordingly.

wennj replied:
I suspected something is wrong, but thought the output names were because of the planemo (ptdk). Anyway, I've adjusted everything and it's much more logical now.

wm75 commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+>    - Click Run Tool
+>
+>    > <comment-title> An easier-to-read table </comment-title>
+>    > This tool creates a table from the VCF file. The columns are tab-deliminated. Take a look at table and see if it is now easier to read. It can also be imported to R/RStudio/Excel more easily.
```
```suggestion
>    > This tool creates a table from the VCF file. The columns are tab-delimited. Take a look at the table and see if it is now easier to read. It can also be imported to R/RStudio/Excel more easily.
```

wm75 commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+
+# SNV Specificity Determination
+
+Now we come to the last but most complex section of this tutorial - determining SNV specificities. We have seen that CpGV-V15 is a mixed isolate. Now we want to find out which isolates were mixed and how this SNV pattern can be explained. In the next step, we will consider  SNV positions as markers and look for SNV positions that are only variable for one or more sequenced isolates of CpGV. If an ...
```
```suggestion
Now we come to the last but most complex section of this tutorial - determining SNV specificities. We have seen that CpGV-V15 is a mixed isolate. Now we want to find out which isolates were mixed and how this SNV pattern can be explained. In the next step, we will consider  SNV positions as markers and look for SNV positions that are only variable for one or more sequenced isolates of CpGV. If a SNV position is only variable for CpGV-S (i.e. the relative frequency is greater than 0; `REL.ALT > 0`) but is equal to zero for the other isolates, then this position is a marker for CpGV-S, or specific for CpGV-S. Since we cannot perform this determination by hand for all positions, I have written a small program implemented in the **Text reformatting with awk** tool that does the work for us. 
```

wm75 commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+>    - *"Numbers of header ines to skip"*: `1`
+>    - Click Run Tool
+>
+>    > <question-title>Mainly two nucleotides only in vairable SNV positions?</question-title>
+>    > Think about an option how the occurance of one, two, three or four nucleotides per SNV position can be analyzed.   
+>    {: .comment}
```
Can you improve the wording of  this question? It's not easy to understand.

wennj replied:
The question was indeed poorly worded. I have revised it complety. Now the reader should not be completely lost.

wm75 reviewed (commented):
Very nice work @wennj !
Those awk expressions are  rather intimidating for beginners I guess, but also elegant compared to multi-step solutions with other tools.
Can be interesting to  teach them :-)

wm75 commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+    SNV: Single nucleotide variant
+    SRA: Sequence Read Archive
+key_points:
+- Baculovirus populations are heterogeneous and show genetic variability.
```
You are mostly talking about isolates, but sometimes also about populations or samples.
I realize the terminology around the concept may be the source of endless debates, but maybe you could reserve a small paragraph in the introduction for explaining what an isolate is in the context here and why it can be a mix of things.

wennj replied:
That is indeed a very important and correct critic, which I have of course addressed. The use of sample and isolate arose from the fact that in the VCF file the columns with the data sets are called ‘sample columns’ (SAMPLE). In an earlier version, I have tried to take this into account (which then led to chaos). I have now consistently used the correct nomenclature isolate and also adapted all steps, outputs, codes and the workflow to it. Population is now defined in the introduction but is no longer used so much in the text.

wennj commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,851 @@
+- Transform the output (VCF file) to a readable table format
+- Interpret the SNV data to analyse the intra-isolate variability of baculovirus isolates or samples.
+- Learn to create your own sequence data and analyse it using the provided workflow and tool to explore the genetic variation of baculovirus populations.
+time_estimation: 2H
```
My student, a total beginner did it in about 2 hours. Changed time estimation value to 3H. Should be a solid and optimistic estimate.

wennj commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.bib:
```diff
@@ -0,0 +1,147 @@
+#
+# You can remove the examples below
+
+@article{Batut2018,
+  doi = {10.1016/j.cels.2018.05.012},
+  url = {https://doi.org/10.1016/j.cels.2018.05.012},
+  year = {2018},
+  month = jun,
+  publisher = {Elsevier {BV}},
+  volume = {6},
+  number = {6},
+  pages = {752--758.e1},
+  author = {B{\'{e}}r{\'{e}}nice Batut and Saskia Hiltemann and Andrea Bagnacani and Dannon Baker and Vivek Bhardwaj and
+           Clemens Blank and Anthony Bretaudeau and Loraine Brillet-Gu{\'{e}}guen and Martin {\v{C}}ech and John Chilton
+           and Dave Clements and Olivia Doppelt-Azeroual and Anika Erxleben and Mallory Ann Freeberg and Simon Gladman and
+           Youri Hoogstrate and Hans-Rudolf Hotz and Torsten Houwaart and Pratik Jagtap and Delphine Larivi{\`{e}}re and
+           Gildas Le Corguill{\'{e}} and Thomas Manke and Fabien Mareuil and Fidel Ram{\'{i}}rez and Devon Ryan and
+           Florian Christoph Sigloch and Nicola Soranzo and Joachim Wolff and Pavankumar Videm and Markus Wolfien and
+           Aisanjiang Wubuli and Dilmurat Yusuf and James Taylor and Rolf Backofen and Anton Nekrutenko and Bj\"{o}rn Gr\"{u}ning},
+  title = {Community-Driven Data Analysis Training for Biology},
+  journal = {Cell Systems}
+}
```
It seems that I have completely messed up the bib file. Quotes have been revised, updated and sorted.

wennj commented:
> Very nice work @wennj ! Those awk expressions are rather intimidating for beginners I guess, but also elegant compared to multi-step solutions with other tools. Can be interesting to teach them :-)

Thank you for your time, the valuable comments and suggestions @wm75 . It took a while to make all the necessary changes but I think that all the comments have been addressed. The tutorial has improved!

wm75 reviewed (commented):
The remaining CI issues are:
```
{"message":"This workflow is missing a test, which is now mandatory. Please see [the FAQ on how to add tests to your workflows](https://training.galaxyproject.org/training-material/faqs/gtn/gtn_workflow_testing.html).","location":{"path":"./topics/variant-analysis/tutorials/baculovirus-isolate-variation/workflows/main_workflow.ga","range":{"start":{"line":1,"column":1},"end":{"line":1,"column":1}}},"severity":"ERROR","code":{"value":"GTN:027","url":"https://training.galaxyproject.org/training-material/gtn_rdoc/Gtn/Linter.html#method-c-"}}
{"message":"This looks like a DOI which could be better served by using the built-in Citations mechanism. You can use https://doi2bib.org/ to convert your DOI into a .bib formatted entry, and add to your tutorial.md","location":{"path":"./topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md","range":{"start":{"line":202,"column":97},"end":{"line":202,"column":159}}},"severity":"WARNING","code":{"value":"GTN:004","url":"https://training.galaxyproject.org/training-material/gtn_rdoc/Gtn/Linter.html#method-c-check_dois"},"suggestions":[{"text":"{% cite ... %}","range":{"start":{"line":202,"column":97},"end":{"line":202,"column":159}}}]}
{"message":"Please do not include an # Introduction section, it is unnecessary here, just start directly into your text. The first paragraph that is seen by our infrastructure will automatically be shown in a few places as an abstract.","location":{"path":"./topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md","range":{"start":{"line":1,"column":1},"end":{"line":2,"column":1}}},"severity":"ERROR","code":{"value":"GTN:046","url":"https://training.galaxyproject.org/training-material/gtn_rdoc/Gtn/Linter.html#method-c-useless_intro"},"suggestions":[{"text":"","range":{"start":{"line":1,"column":1},"end":{"line":2,"column":1}}}]}
```

Let me know if  you have trouble with the workflow tests ...

... and please remember to pull my changes first before continuing to work on your local version.

wennj commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,870 @@
+> |:------------------:|:------------------:|:------------------:|:------------------:|
+> | CpGV-M | KM217575 | [SRR31589148](https://trace.ncbi.nlm.nih.gov/Traces?run=SRR31589148) | {% cite Wennmann2020 %}
+> | CpGV-S | KM217573 | [SRR31589147](https://trace.ncbi.nlm.nih.gov/Traces?run=SRR31589147) | {% cite Wennmann2020 %} |
+> | CpGV-E2 | KM217577 | [SRR31589146](https://trace.ncbi.nlm.nih.gov/Traces?run=SRR31589146) | [Gueli Alletti et al. 2017](https://doi.org/10.3390/v9090250) |
```
Overlooked missing citation added.

wennj commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,870 @@
+- Determine variable SNV positions for multiple isolates using a common reference genome.
+- Transform the output (VCF file) to a readable table format.
+- Interpret the SNV data to analyse the intra-isolate variability of baculoviruses.
+- Learn to create your own sequence data and analyse it using the provided workflow and tool to explore the genetic variation.
```
Statement removed.

wennj commented:
@wm75 I think I pulled your changes this time. And I'm starting to understand how the automated tests work. I have added the workflow test files, as well. Let me know if any changes are required. Sorry for the trouble with the Gemfile.lock.

wm75 commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/workflows/Workflow-to-detect-SNV-from-Illumina-sequenced-baculovirus-isolates-tests.yml:
```diff
@@ -0,0 +1,11 @@
+- doc: Test outline for Workflow-to-detect-SNV-from-Illumina-sequenced-baculovirus-isolates
+  job: {}
+  outputs:
+    bcftools_call_output:
+      path: test-data/bcftools_call_output.vcf
+    vcf_table_ALT1:
+      path: test-data/vcf_table_ALT1.tabular
+    vcf_table_ALT1_specificity:
+      path: test-data/vcf_table_ALT1_specificity.tabular
+    vcf_table:
+      path: test-data/vcf_table.tabular
```
I'm ok with leaving the workflow test as a template for now until we have more experience with this tutorial.
Quite likely we will want to address some of these issues in the future:
- do we really want to download original, big NCBI datasets or would it be better to do what most tutorials do and use trimmed down sample data deposited on zenodo?
- can we split the variant report by sample in a more scalable way (by splitting the table based on the isolate column into a collection, for example)
- can the awk specificity script be generalized, e.g. by first generating all possible specificity groups from which we then subtract user-provided ones that we don't want to plot, etc.

Things like that will make the workflow much more generally useful instead of just letting you reanalyze hardcoded datasets.

What this also means, however, is that the WF would likely change a lot during this process and so it makes little sense to test for exact identity of outputs and include those  result files in the PR.
Instead I would recommend checking that specific text appears in the outputs, which indicates WF success and which you wouldn't expect to change much with minor modifications to the anaylsis.
https://training.galaxyproject.org/training-material/topics/contributing/tutorials/create-new-tutorial-technical/tutorial.html#testing-the-workflow-recommended demonstrates how to do this.

wennj replied:
Oh ok, that's indeed a good idea, which actually didn't occur to me before. But I would like to implement your ideas:
- split the variant report by sample into a collection, and 
- determine all specificities and let the user decide which one to plot. (Even if I don't know yet how to incorporate this decision into the workflow).

Indeed, this will make the workflow less hardcoded(?) but more flexilbe.

I see the advantage of using zenodo datasets, but in my workflow, all the NCBI datasets loaded are only about one GB and the delay is minimal. I also think that from time to time, students should learn how to load data from NCBI (using galaxy) and work with the obtained collections. But I totally understand your point.
 
But aren't the changes too much for this PR? How do we proceed?

wennj replied:
And I now understand the points around the workflow tests. I'll adjust the test when I know how we're going to proceed.

wm75 replied:
Sorry for the delay. Been sick last week.
I also thought that we should get a first version of this merged as soon as possible and improve things later on.
For now I would just ask you to change the WF tests because it makes little sense to add the test datasets now and remove them with the next iteration again.
Then we can merge, but you are, of course, more than welcome to implement any of the above suggestions later or improve things in other ways. GTN tutorials are far from contribute once and never come back static artifacts ;-)

wm75 commented on topics/variant-analysis/tutorials/baculovirus-isolate-variation/tutorial.md:
```diff
@@ -0,0 +1,867 @@
+>
+>    > <comment-title> Filtering the CpGV-S data </comment-title>
+>    >
+>    > The VCF SNV table is reduced to the data of CpGV-S. This is done by filtering on column 23 (ISOLATE).
```
Just noticed that this comment talks about column 23, but you're configuring c25 in the hands-on.

wennj commented:
@wm75 Sorry also from my side for the late feedback. Very busy this week. 
I am a bit confused now what to do with the WF tests. 1. I can leave them as they are for now, although they make few sense and add changes as suggested by you later. 2. I remove the WF tests entirely for now and add updated WF tests later. 3. I add your suggested changed to the WF tests which would require significant changes.

wm75 commented:
> @wm75 Sorry also from my side for the late feedback. Very busy this week. I am a bit confused now what to do with the WF tests. 1. I can leave them as they are for now, although they make few sense and add changes as suggested by you later. 2. I remove the WF tests entirely for now and add updated WF tests later. 3. I add your suggested changed to the WF tests which would require significant changes.

Let me see if I can come up with something this week :-)

wennj commented:
> > @wm75 Sorry also from my side for the late feedback. Very busy this week. I am a bit confused now what to do with the WF tests. 1. I can leave them as they are for now, although they make few sense and add changes as suggested by you later. 2. I remove the WF tests entirely for now and add updated WF tests later. 3. I add your suggested changed to the WF tests which would require significant changes.
> 
> Let me see if I can come up with something this week :-)

Do you already have an idea? I am motivated to continue working on the tutorial, even though it is merged.

wm75 reviewed (approved):
I changed the tests now to confirm that expected content is present in the workflow outputs and removed the test-data files (which we would likely have had to update on any future change).

Lets get this live and see how it goes :-)
